strawberry/starlite/controller.py

```diff
@@ -0,0 +1,399 @@
+"""Starlite integration for strawberry-graphql."""
+import json
+from datetime import timedelta
+from typing import TYPE_CHECKING, Any, Dict, Iterable, List, Optional, Type, Union, cast
+
+from pydantic import BaseModel
+
+from starlite import (
+    BackgroundTasks,
+    Controller,
+    MediaType,
+    Provide,
+    Request,
+    Response,
+    ResponseSpec,
+    WebSocket,
+    get,
+    post,
+    websocket,
+)
+from starlite.exceptions import ImproperlyConfiguredException
+from starlite.status_codes import (
+    HTTP_200_OK,
+    HTTP_400_BAD_REQUEST,
+    HTTP_404_NOT_FOUND,
+    HTTP_415_UNSUPPORTED_MEDIA_TYPE,
+)
+from strawberry.exceptions import InvalidCustomContext, MissingQueryError
+from strawberry.file_uploads.utils import replace_placeholders_with_files
+from strawberry.http import (
+    GraphQLHTTPResponse,
+    parse_query_params,
+    parse_request_data,
+    process_result,
+)
+from strawberry.schema import BaseSchema
+from strawberry.schema.exceptions import InvalidOperationTypeError
+from strawberry.subscriptions import GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL
+from strawberry.types import ExecutionResult
+from strawberry.types.graphql import OperationType
+from strawberry.utils.debug import pretty_print_graphql_operation
+from strawberry.utils.graphiql import get_graphiql_html
+
+from .handlers.graphql_transport_ws_handler import (
+    GraphQLTransportWSHandler as BaseGraphQLTransportWSHandler,
+)
+from .handlers.graphql_ws_handler import GraphQLWSHandler as BaseGraphQLWSHandler
+
+
+if TYPE_CHECKING:
+    from starlite.types import Dependencies
+
+
+CustomContext = Union["BaseContext", Dict[str, Any]]
+MergedContext = Union[
+    "BaseContext", Dict[str, Union[Any, BackgroundTasks, Request, Response, WebSocket]]
+]
+
+
+async def _context_getter(
+    custom_context: Optional[CustomContext],
+    request: Request,
+) -> MergedContext:
+    if isinstance(custom_context, BaseContext):
+        custom_context.request = request
+        return custom_context
+    default_context = {
+        "request": request,
+    }
+    if isinstance(custom_context, dict):
+        return {
+            **default_context,
+            **custom_context,
+        }
+    if custom_context is None:
+        return default_context
+    raise InvalidCustomContext()
+
+
+class GraphQLResource(BaseModel):
+    data: Optional[Dict[str, object]]
+    errors: Optional[List[object]]
+    extensions: Optional[Dict[str, object]]
+
+
+class EmptyResponseModel(BaseModel):
```

Comment:
For optimal performance it's recommended to use dataclasses.  Pydantic models are significantly slower to serialize and deserialize.